### PR TITLE
Support for unicode strings

### PR DESCRIPTION
**Pull request by isubasinghe**
This PR adds support for char's and consequently unicode. Unfortunately, this means we cannot talk about chars as bytes anymore even when it makes sense to do so, this is to be addressed in a following PR.

### CODE REVIEW

File: source/pervasive/string.rs [hunk 1]
```diff
@@ -25,47 +25,82 @@ pub const fn new_strlit<'a>(s: &'a str) -> StrSlice<'a> {
...
 
+    pub spec fn is_ascii(&self) -> bool;
+    
+    // len returns the byte length, this returns
+    // the unicode length 
+    // Warning: O(n)
     #[verifier(external_body)]
-    pub fn len(&self) -> (l: usize)
-        ensures self.is_ascii() ==> l as nat === self.bytes().len()
+    pub fn unicode_len(&self) -> (l: usize)
```
[the remark this answers is not in the document]
Comment (utaal) [reply source/pervasive/string.rs:36]: I think this PR comment should be a docstring for the function instead.

File: source/pervasive/string.rs [hunk 2]
```diff
@@ -25,47 +25,83 @@ pub const fn new_strlit<'a>(s: &'a str) -> StrSlice<'a> {
...
+            else { break; }
+        }
+        let byte_start = byte_start.unwrap();
+        let byte_end = byte_end.unwrap();
+        StrSlice {
+            inner: &self.inner[byte_start..byte_end]
+        }
+    }
+
     #[rustc_diagnostic_item = "pervasive::string::StrSlice::to_string"]
```
Comment (utaal) [inline source/pervasive/string.rs:100]: This isn't in this pull request, but why do you need this diagnostic_item?

File: source/pervasive/string.rs [hunk 3]
```diff
@@ -25,47 +25,83 @@ pub const fn new_strlit<'a>(s: &'a str) -> StrSlice<'a> {
...
+    /// The len() function in rust returns the byte length. 
+    /// It is more useful to talk about the length of characters and therefore this function was added. 
+    /// Please note that this function counts the unicode variation selectors as characters. 
+    /// Warning: O(n)
     #[verifier(external_body)]
-    pub fn len(&self) -> (l: usize)
-        ensures self.is_ascii() ==> l as nat === self.bytes().len()
+    pub fn unicode_len(&self) -> (l: usize)
+        ensures 
+            l as nat === self.view().len()
```
Comment (utaal) [inline source/pervasive/string.rs:39]: You can use `@` for .view() everywhere except for the declaration.

File: source/pervasive/string.rs [hunk 4]
```diff
@@ -25,47 +25,83 @@ pub const fn new_strlit<'a>(s: &'a str) -> StrSlice<'a> {
...
+        }
+        let byte_start = byte_start.unwrap();
+        let byte_end = byte_end.unwrap();
+        StrSlice {
+            inner: &self.inner[byte_start..byte_end]
+        }
+    }
+
     #[rustc_diagnostic_item = "pervasive::string::StrSlice::to_string"]
     #[verifier(external_body)]
```
Comment (utaal) [inline source/pervasive/string.rs:101]: I didn't notice this before, but I don't think this needs to be `external_body` as it calls a function with a requires/ensures.

File: source/rust_verify/tests/strings.rs [hunk 5]
```diff
@@ -267,3 +266,156 @@ test_verify_one_file! {
...
+        use pervasive::string::*;
+        fn test() {
+            let a = new_strlit("ABC");
+            let b = a.to_string();
+            let c = b.as_str();
+            reveal_strlit("ABC");
+            assert(a@.ext_equal(c@));
+            assert(a.is_ascii());
+            // TODO assert(b.is_ascii());
+            // TODO assert(c.is_ascii());
```
Comment (utaal) [inline source/rust_verify/tests/strings.rs:418]: I believe these should verify. I think some of the pervasive::string functions need to maintain/forward information about `is_ascii`.

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/pervasive/string.rs (hunk 2 above)
Comment (isubasinghe) [reply source/pervasive/string.rs:100]: removed

File: source/rust_verify/example/playground.rs [hunk 6]
```diff
@@ -1,37 +1,25 @@
-use builtin_macros::*;
```
Comment (utaal) [inline source/rust_verify/example/playground.rs:1]: Please remove this change.

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/rust_verify/example/playground.rs (hunk 6 above)
Comment (isubasinghe) [reply source/rust_verify/example/playground.rs]: Sorry I should really avoid using `git add .`

**Review by utaal — APPROVED** [review]

**Pull request merged**